### PR TITLE
Angular Security Best Practices

### PR DESCRIPTION
**Pull request by fredericghijselinck**

### CODE REVIEW

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+
+## 1. Up-to-date Angular libraries
+
+The angular team is doing releases at regular intervals for feature enhancements, bug fixes and security patches as appropriate.
+So, it is recommended to update the Angular libraries at regular intervals.
+Not doing so may allow attackers to attack the app using known security vulnerabilities present with older releases.
+
+
+# Preventing cross-site scripting (XSS)
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: Why is this a `h1`? Breaks the layout with huge margin on top

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+> Software security is a hot topic nowadays.
+We, web developers, need to be up-to-date with all latest security issues that we could encounter when developing a web application.
+In this blog we'll check what kind of best practices we should have in mind when building an Angular app so we limit the amount of security issues we could have.
+
+
+## 1. Up-to-date Angular libraries
+
+The angular team is doing releases at regular intervals for feature enhancements, bug fixes and security patches as appropriate.
+So, it is recommended to update the Angular libraries at regular intervals.
+Not doing so may allow attackers to attack the app using known security vulnerabilities present with older releases.
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: present within older releases?

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  exports: [CommonModule, ApplicationModule]
+})
+export class BrowserModule {}
+{% endhighlight %}
+
+
+### The DOM sanitization service
+The goal of the DomSanitizer is to clean untrusted parts of values.
+
+The skeleton of the class looks like:
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: ... looks like this?

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  abstract bypassSecurityTrustResourceUrl(value: string): SafeResourceUrl;
+}
+{% endhighlight %}
+
+As you can see, there are two kinds of method patterns.
+The first one is the `sanitize` method, which gets the _context_ and an _untrusted value_ and returns a trusted value.
+The other ones are the `bypassSecurityTrustX` methods which are getting the _untrusted value_ according to the value usage and are returning a trusted object.
+
+#### The sanitize method
+If a value is trusted for the context, this sanitize method will (in case of a SafeValue) unwrap the contained safe value and use it directly.
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: `SafeValue` in code tag?

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+Otherwise, the value will be sanitized to be safe according to the security context.
+
+There are three main helper functions for sanitizing the values.
+The `sanitizeHtml` function sanitizes the untrusted HTML value by parsing the value and checks its tokens.
+The `sanitizeStyle` and `sanitizeUrl` functions sanitize the untrusted style or URL value by regular expressions.
+
+#### How can we disable the sanitization logic?
+In specific situations, it might be necessary to disable sanitization.
+Users can bypass security by constructing a value with one of the `bypassSecurityTrustX` methods, and then binding to that value from the template.
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: Maybe add an example here?

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+> Be careful: If you trust a value that might be malicious, you are introducing a security vulnerability into your application!
+
+
+## 3. Content security policy (CSP)
+
+Content Security Policy is a defense-in-depth technique to prevent XSS.
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: The MDN source is also pretty good, with nice introduction. https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP 

Content Security Policy (CSP) is an added layer of security that helps to detect and mitigate certain types of attacks, including Cross Site Scripting (XSS) and data injection attacks. These attacks are used for everything from data theft to site defacement or distribution of malware.

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+To enable CSP, configure your web server to return an appropriate `Content-Security-Policy` HTTP header.
+More info about specifying your CSP can be found on [Google Developers Page about Content Security Policy](https://developers.google.com/web/fundamentals/security/csp/){:target="blank" rel="noopener noreferrer"}.
+To check if your CSP is valid you can use the [CSP evaluator from google](https://csp-evaluator.withgoogle.com){:target="blank" rel="noopener noreferrer"}.
+
+
+## 4. Use the offline template compiler
+
+_Angular templates_ are the same as executable code: HTML, attributes, and binding expressions (but not the values bound) in templates are trusted to be safe.
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: Offline compiler or AOT compiler? https://angular.io/guide/aot-compiler

Maybe add info how easy it is to enable --aot with angular cli

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+Injecting template code into an Angular application is the same as injecting executable code into the application.
+So, validate all data on server-side code and escape appropriately to prevent XSS vulnerabilities on the server.
+Also, Angular recommends _not to generate Angular templates on the server side_ using a templating language.
+
+
+# HTTP-level vulnerabilities
+
+Angular has built-in support to help prevent two common HTTP vulnerabilities, cross-site request forgery (CSRF or XSRF) and cross-site script inclusion (XSSI).
+Both of these must be mitigated primarily on the server side, but Angular provides helpers to make integration on the client side easier.
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: How do they do that?

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+
+## 6. Server-side XSS protection
+
+Injecting template code into an Angular application is the same as injecting executable code into the application.
+So, validate all data on server-side code and escape appropriately to prevent XSS vulnerabilities on the server.
+Also, Angular recommends _not to generate Angular templates on the server side_ using a templating language.
+
+
+# HTTP-level vulnerabilities
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: This is again a `h1` but list continues on index 7?

File: _posts/2018-03-30-angular-security-best-practices.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+## 8. Cross-site script inclusion (XSSI)
+
+Cross-site script inclusion (also known as **JSON vulnerability**) can allow an attacker's website to read data from a JSON API.
+The attack works on older browsers by overriding native JavaScript object constructors, and then including an API URL using a <script> tag.
+This attack is only successful if the returned JSON is executable as JavaScript.
+
+Servers can prevent an attack by prefixing all JSON responses to make them non-executable, by convention, using the well-known string `")]}',\n"`.
+Angular's HttpClient library recognizes this convention and automatically strips the string `")]}',\n"` from all responses before further parsing.
+
```
Comment (MrJean) [inline _posts/2018-03-30-angular-security-best-practices.md]: Do we need a conclusion here?

File: _posts/2018-03-30-angular-security-best-practices.md (hunk 8 above)
Comment (fredericghijselinck) [reply _posts/2018-03-30-angular-security-best-practices.md]: That is explained in the next 2 chapters

**Pull request merged**